### PR TITLE
Chainlink integration

### PR DESCRIPTION
**Pull request by piobab**
Astroport Anchored Chainlink (normalized to UST):

1. System queries the Chainlink price.
    1. If the Chainlink price is USD denominated, the system converts the price to UST using the Chainlink UST/USD price feed. The remaining UST denominated price is `Chainlink UST Price`.
2. System queries the `Astroport TWAP`.
3. System checks whether the `Astroport TWAP` falls within the `max_deviation` from the `Chainlink UST Price`. Specifically, the `Astroport TWAP` should fall within the following interval: `[Chainlink UST Price * (1 - max_deviation), Chainlink UST Price * (1 + max_deviation)]`
    1. If it doesn’t, the system is halted and the transaction doesn’t go through.
4. `max_deviation` = 15% initially for all pools, but could be adjusted by governance.

Chainlink Solo:

1. System queries the Chainlink price.
    1. If the Chainlink price is USD denominated, the system converts the price to UST using the Chainlink UST/USD price feed. The remaining UST denominated price is `Chainlink UST Price`.

### CODE REVIEW

**Review by dancreee — APPROVED** [review]

File: contracts/mars-oracle/Cargo.toml [hunk 1]
```diff
@@ -32,12 +32,14 @@ cw-storage-plus = "0.9.0"
 cosmwasm-std = "0.16.2"
 
 schemars = "0.8.1"
-serde = { version = "1.0.103", default-features = false, features = ["derive"] }
+serde = { version = "1.0.136", default-features = false, features = ["derive"] }
```
Comment (larry0x) [inline contracts/mars-oracle/Cargo.toml:35]: any specific reason to update this dep?

i don't think changing the minor version number will break anything, but we should make deps version numbers consistent across all contracts (e.g. just use `1.0` without specifying the 3rd position.)

Comment (piobab) [reply contracts/mars-oracle/Cargo.toml:35]: Yes, there was dependency issue with chainlink before (with ver 0.1.0 as I remember) but currently with chainlink ver 1.0.0 I see that everything is ok so I leave prev ver.

Agree with you that we could use 1.0 in contracts.

File: contracts/mars-oracle/src/contract.rs [hunk 2]
```diff
@@ -97,6 +100,12 @@ pub fn execute_set_asset(
         | PriceSourceChecked::AstroportTwap { pair_address, .. } => {
             assert_astroport_pool_assets(&deps.querier, &asset, pair_address)?;
         }
+        PriceSourceChecked::ChainlinkAnchoredToAstroportTwap {
+            proxy_address: _,
+            ust_usd_proxy_address: _,
+            pair_address,
+            ..
+        } => assert_astroport_pool_assets(&deps.querier, &asset, pair_address)?,
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs]: Thx, it was broken after applying updates.

File: contracts/mars-oracle/src/contract.rs [hunk 3]
```diff
@@ -128,6 +137,14 @@ pub fn execute_record_twap_snapshots(
...
                 tolerance,
             } => (pair_address, window_size, tolerance),
+            PriceSourceChecked::ChainlinkAnchoredToAstroportTwap {
+                proxy_address: _,
+                ust_usd_proxy_address: _,
+                pair_address,
+                window_size,
+                tolerance,
+                deviation_percentage: _,
+            } => (pair_address, window_size, tolerance),
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs]: Fixed

File: contracts/mars-oracle/src/error.rs [hunk 4]
```diff
@@ -24,6 +24,9 @@ pub enum ContractError {
 
     #[error("Invalid pair")]
     InvalidPair {},
+
+    #[error("Chainlink price exceeds the allowable price deviation from Astroport TWAP")]
+    ChainlinkPriceExceedsAllowableDeviationPercentage {},
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/error.rs]: Yes, added chainlink price, astroport twap and applied deviation percentage.

File: contracts/mars-oracle/src/contract.rs [hunk 5]
```diff
@@ -315,6 +301,132 @@ fn query_asset_price(
...
+
+            let chainlink_price = query_chainlink_spot_price(
+                &deps.querier,
+                &proxy_address,
+                ust_usd_proxy_address.as_ref(),
+            )?;
+
+            // Check whether the Astroport TWAP price falls within the max_deviation from the Chainlink price
+            let deviation_percentage =
+                deviation_percentage.unwrap_or_else(|| Decimal::from_ratio(15u32, 100u32));
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs]: Good point. Fixed.

File: contracts/mars-oracle/src/contract.rs [hunk 6]
```diff
@@ -315,6 +301,132 @@ fn query_asset_price(
...
+        let ust_usd_decimals: u8 =
+            querier.query_wasm_smart(ust_usd_proxy_address, &ChainlinkQueryMsg::Decimals)?;
+        let ust_usd_round: Round =
+            querier.query_wasm_smart(ust_usd_proxy_address, &ChainlinkQueryMsg::LatestRoundData)?;
+
+        let ust_usd_price_precision = Uint128::from(10_u128.pow(ust_usd_decimals.into()));
+        let ust_usd_price =
+            Decimal::from_ratio(ust_usd_round.answer as u128, ust_usd_price_precision);
+
+        let price = price.checked_div(ust_usd_price)?;
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs]: I made a little refactoring.

File: contracts/mars-oracle/src/contract.rs [hunk 7]
```diff
@@ -315,6 +301,132 @@ fn query_asset_price(
...
+
+        Ok(price)
+    } else {
+        Ok(price)
```
[the remark this answers is not in the document]
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs]: 'next to each other' :D Updated ;)

**Review by larry0x — APPROVED** [review]

**Pull request closed without merging**